### ISSUE TITLE
User enumeration of file-based backend based on timing attack

### ISSUE THREAD

**Issue by TheHllm** [authelia/authelia#949]
It is possible to tell from the outside if a username is valid, by simply performing a login attempt.
If the server responds quickly to the request the username is invalid, if however the server dose not respond quickly the password is at fault.

To fix this issue the server should take a fixed minimum amount of time to respond to login requests.
This would have to be implemented in a way, such that the server only takes (significantly longer) for a valid login.

In pseudo-code:
```
int fixedTime = 10; // should be configurable so that this time is grater than the time it takes to hash a password
 int login(u,p){
 if(username_valid(u)){
  string hash = getHash(u);
  if(checkHash(hash, pw)){
   //do some other login stuff that may take longer that a wait would have.....
   return 200;
  }else{
   waitForRemaingTime(Date.now, time);
   return 403;
  }
 }else{
   waitForRemaingTime(Date.now, time);
   return 403;
 }
}
```

**Comment by james-d-elliott** [comment]
Hey @TheHllm thanks for reporting this. Is my understanding correct that you're using the file based authentication backend?

**Comment by TheHllm** [comment]
@james-d-elliott  Yes

**Comment by james-d-elliott** [comment]
Alright. So the reason this is happening most likely, if the user doesn't exist we don't has the input password to compare it. I will discuss it with the other maintainers and see what they think. While it's good for performance it slightly reduces security.

To put your mind at ease, [regulation](https://docs.authelia.com/configuration/regulation.html) can be used to block brute force attacks.

**Comment by TheHllm** [comment]
The relevant code is (as far as i can tell) in [handler_firstfactor.go](https://github.com/authelia/authelia/blob/master/internal/handlers/handler_firstfactor.go) but i have never done any go....
It should only impact performance for invalid logins.
And i agree with you that brute force is unrealisic if you use regulation

**Comment by james-d-elliott** [comment]
Related due to inability to solve this: #478

**Comment by clems4ever** [comment]
> Alright. So the reason this is happening most likely, if the user doesn't exist we don't has the input password to compare it. I will discuss it with the other maintainers and see what they think. While it's good for performance it slightly reduces security.
> 
> While [regulation](https://docs.authelia.com/configuration/regulation.html) can be used to block brute force attacks, if they're looking for usernames.. this would get them.

After discussing with @james-d-elliott, regulation does not really help preventing enumeration because it bans accounts and not IPs. And as mentioned in the previous comment, #478 might indeed prevent someone from using fail2ban to ban IPs.

I think @TheHllm's proposal makes real sense, thanks for reporting btw. We will fix that in the next coming hours. Also I think it might be worth having this mechanism for any kind of backend, what do you think @james-d-elliott?

**Comment by clems4ever** [comment]
Here is a simple update to @TheHllm's proposal because it might be complicated setting `fixedTime` manually for an admin. What we can do is execute some hashing at startup to measure an average value (so it's based on actual target machine performance) and make sure the time elapsed by /api/firstfactor endpoit is always the same fixed value (in order for multiple instances of Authelia running on various machines to always take the same amount of time) taking into account the time it takes to hash on average in case the user does not exist and by compensating in other cases.

What do you guys think?

**Comment by james-d-elliott** [comment]
I agree. I think it's important to fix it for the file provider at the very least and we can add it to the user provider at a later of necessary.

**Comment by clems4ever** [comment]
@TheHllm , thank you very much for reporting that issue. I see it as a great contribution and you'll be mentioned in the repo for sure. Next time I would recommend ,for your own safety and for everyone's safety, to contact one of the core contributors first on matrix or Gitter or by email (https://github.com/authelia/authelia#security) before creating a github issue so that we can disclose the issue once the fix is ready.
No hard feeling whatsoever, just a recommendation for everyone's safety. Thank you very much again. We'll keep you posted on the fix.

**Comment by james-d-elliott** [comment]
@TheHllm hotfix has been released in 4.16.0, we're going to implement a long term fix in the next few days which will effectively close this issue. We will probably organize a security advisory to be posted. Once both of these things are done this issue will be closed.

Thanks so much for reporting. We hope you and other users continue to let us know about security flaws. Ideally if it's done privately so we can provide a fix before someone exploits it as Clement mentioned, but we'd rather know than not!

**Comment by TheHllm** [comment]
> Here is a simple update to @TheHllm's proposal because it might be complicated setting `fixedTime` manually for an admin. What we can do is execute some hashing at startup to measure an average value (so it's based on actual target machine performance) and make sure the time elapsed by /api/firstfactor endpoit is always the same fixed value (in order for multiple instances of Authelia running on various machines to always take the same amount of time) taking into account the time it takes to hash on average in case the user does not exist and by compensating in other cases.
> 
> What do you guys think?

I agree that would be a better solution but depending on the hashing-ago this may not be so simple.
Argon for example has, like many others (e.g. bcrypt), cost parameter(s). This means that the time it takes to verify one password against a hash might not be the same as against an other hash.
This would mean that the hash used for generating this time would have to be the one (out of all the hashes / potentially other auth methods) that takes the longest to complete. This would mean finding this hash wich will take some time.
Of course in most cases all hashes will have the same cost, just something to keep in mind.

**Comment by james-d-elliott** [comment]
That is a definite possibility. However we can put clear warnings about it. This should only happen when people reduce the cost of the hashing algorithm without forcing all users to reset their password.

Let me explain. If the settings current time cost is 2.1s, and users have a weaker time cost hash, of 1.5s, as long as we make sure the response to 1FA login only happens at 2.1s, all login times  are the same.

We can additionally increase the consistency of this automatically for users by updating the hash in the database (YAML in this case) when a successful login occurs and the hash settings differ. 

So I propose the following stages:
1. Our current patch which forces hashing on all password inputs regardless.
2. Implement startup check of current settings time cost and document harmful consequences of reducing the hash settings without requiring password resets.
3. Consider updating the hash in user storage.

**Comment by james-d-elliott** [comment]
@TheHllm the above PR solves this long term and will be released with v4.19.0. It adds a automatically adjusting delay to all auth providers padded with a 0-85ms random delay.
